kore/src/Kore/Syntax/Id.hs

```diff
@@ -47,7 +47,7 @@ import System.IO.Unsafe (unsafePerformIO)
 
 data InternedTextCache = InternedTextCache
     { counter :: {-# UNPACK #-} !Word
-    , internedTexts :: !(HashMap Text Word)
+    , internedTexts :: !(HashMap Text InternedText)
```

Comment:
Does it increase the size of serialized data?

Reply:
Haven't measured it but I think it should only increase the hashmap size by 1 pointer per entry.

The reason being: the `Text` object that the entry's key points to, is the same `Text` object that the `InternedText` value points to:

```hs
internText text =
  ...
                    HashMap.alterF
                        \case
                            ...
                            -- Otherwise, create a new ID for it and intern it.
                            Nothing ->
                                let newIden = counter
                                    newInterned = InternedText text newIden
                                    ^^^^ ***the new value, containing a pointer to the same text object as the key***
                                 in ((newInterned, counter + 1), Just newInterned)
                        text
                        ^^^^ ***the new key***
```

So the old memory layout would have looked like this:

```
Entry: Key           Value
        |              |
        |              |
        ↓              ↓
        Text         Word (with an unpacked #Word)
```

And it should now look something like this (notice the extra pointer `InternedText -> Text`:

```
Entry: Key           Value
        |              |
        |              |
        ↓              ↓
        Text <--- InternedText (with an unpacked #Word)
```


On the other hand, wherever an interned string is used, it'll now reuse the same `InternedString` box, instead of creating a new `InternedString` box every time we intern a string. So we'll likely save some memory (and a negligible amount of time) there.

Regarding the serialization itself, it's done using [compact regions](https://hackage.haskell.org/package/compact-0.2.0.0), so the file's data layout should resemble the in-memory layout.

Reply:
Ok, sounds good, thank you for this explanation :sunglasses: